qiskit_optimization/applications/ising/tsp.py

```diff
@@ -1,274 +1,110 @@
-
-    Returns:
-        float: value of the cut.
-    """
-    ret = 0.0
-    for i in range(len(z) - 1):
-        ret += w[z[i], z[i + 1]]
-    ret += w[z[-1], z[0]]
-    return ret
-
-
-def tsp_feasible(x):
-    """Check whether a solution is feasible or not.
-
-    Args:
-        x (numpy.ndarray) : binary string as numpy array.
-
-    Returns:
-        bool: feasible or not.
-    """
-    n = int(np.sqrt(len(x)))
-    y = np.zeros((n, n))
-    for i in range(n):
-        for p__ in range(n):
-            y[i, p__] = x[i * n + p__]
-    for i in range(n):
-        if sum(y[i, p] for p in range(n)) != 1:
-            return False
-    for p__ in range(n):
-        if sum(y[i, p__] for i in range(n)) != 1:
-            return False
-    return True
-
-
-def get_tsp_solution(x):
-    """Get graph solution from binary string.
-
-    Args:
-        x (numpy.ndarray) : binary string as numpy array.
-
-    Returns:
-        list[int]: sequence of cities to traverse.
-            The i-th item in the list corresponds to the city which is visited in the i-th step.
-            The list for an infeasible answer e.g. [[0,1],1,] can be interpreted as
-            visiting [city0 and city1] as the first city, then visit city1 as the second city,
-            then visit no where as the third city).
-    """
-    n = int(np.sqrt(len(x)))
-    z = []
-    for p__ in range(n):
-        p_th_step = []
+    def _build_quadratic_program(self):
+        mdl = Model(name='tsp')
+        n = self._g.number_of_nodes()
+        x = {(i, p): mdl.binary_var(name='x_{0}_{1}'.format(i, p))
+             for i in range(n) for p in range(n)}
+        tsp_func = mdl.sum(self._g.edges[i, j]['weight'] * x[(i, p)] * x[(j, (p+1) % n)]
+                           for i in range(n) for j in range(n) for p in range(n) if i != j)
+        mdl.minimize(tsp_func)
         for i in range(n):
-            if x[i * n + p__] >= 0.999:
-                p_th_step.append(i)
-        if len(p_th_step) == 1:
-            z.extend(p_th_step)
-        else:
-            z.append(p_th_step)
-    return z
+            mdl.add_constraint(mdl.sum(x[(i, p)] for p in range(n)) == 1)
+        for p in range(n):
+            mdl.add_constraint(mdl.sum(x[(i, p)] for i in range(n)) == 1)
+        qp = QuadraticProgram()
+        qp.from_docplex(mdl)
+        return qp
+
+    def plot_graph(self, x, pos=None):
+        route = self.interpret(x)
+        nx.draw(self._g, with_labels=True, pos=pos)
+        nx.draw_networkx_edges(
+            self._g,
+            pos,
+            edgelist=[(route[i], route[(i+1) % len(route)]) for i in range(len(route))],
+            width=8, alpha=0.5, edge_color="tab:red",
+            )
+
+    def interpret(self, x):
+        n = int(np.sqrt(len(x)))
+        route = []
+        for p__ in range(n):
+            for i in range(n):
+                if x[i * n + p__]:
+                    route.append(i)
+        return route
+
+    @staticmethod
```

Comment:
Would this method and the parse_tsplib_format method not be better creating a TSP object with the graph so the user directly gets the TSP application rather than a graph that they then have to turn into a TSP to use at this application level.